### PR TITLE
CP-43777: Add xapi-expiry-alerts library

### PR DESCRIPTION
**Pull request by gangj**

### CODE REVIEW

File: packages/xs-extra/xapi-expiry-alerts.master/opam [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+]
+depends: [
+  "ocaml"
+  "dune" {build & >= "1.4"}
+  "xapi-client"
+  "xapi-consts"
+  "xapi-types"
+  "xapi-stdext-date"
+]
+synopsis: "Library required by xapi"
```
Comment (lindig) [inline packages/xs-extra/xapi-expiry-alerts.master/opam:19]: This needs a better description.

**Review by lindig — CHANGES_REQUESTED** [review]

File: packages/xs-extra/xapi-expiry-alerts.master/opam (hunk 1 above)
Comment (lindig) [reply packages/xs-extra/xapi-expiry-alerts.master/opam:19]: Please also run `opam lint --normalise opam`

**Comment by psafont** [comment]
This is awkward, adding this library will break until the library is in xen-api's master branch.

I recommend creating a PR against xen-api's master branch with
https://github.com/xapi-project/xen-api/pull/5060 and https://github.com/xapi-project/xen-api/pull/5085 (with the commits cleaned up).

After that we can merge this to make the CI happy and be able to use the library in the proprietary daemons

**Comment by gangj** [comment]
>

File: packages/xs-extra/xapi-expiry-alerts.master/opam (hunk 1 above)
Comment (gangj) [reply packages/xs-extra/xapi-expiry-alerts.master/opam]: OK, updated

**Comment by psafont** [comment]
Please squash all three commits

**Comment by gangj** [comment]
> Please squash all three commits

Squashed, please help to review again, thank you~

**Review by lindig — APPROVED** [review]

File: packages/xs-extra/xapi-expiry-alerts.master/opam [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  "xapi-client"
+  "xapi-consts"
+  "xapi-types"
+  "xapi-stdext-date"
+]
+build: [
+  ["dune" "build" "-p" name "-j" jobs]
+  ["dune" "runtest" "-p" name "-j" jobs] {with-test}
+]
+dev-repo: "git+https://github.com/xapi-project/xen-api.git"
```
[the remark this answers is not in the document]
Comment (gangj) [reply packages/xs-extra/xapi-expiry-alerts.master/opam:26]: Sorry, I did not find related spec for the `dev-repo` field, seems it is a string for information purpose only?
https://opam.ocaml.org/doc/Manual.html#File-hierarchies :
```
dev-repo: <string>: the URL of the package's source repository, which may be useful for developers
```

This URL is copied from other opam files:
```
[gangj@xenrt1015818053 xs-opam]$ git grep -wn "git+https://github.com/xapi-project/xen-api.git" packages/xs-extra/
packages/xs-extra/gzip.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/http-lib.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/pciutil.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/rrd-transport.master/opam:5:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/rrd2csv.master/opam:8:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/rrddump.master/opam:11:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/safe-resources.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/sexpr.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/stunnel.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/uuid.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/varstored-guard.master/opam:5:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/vhd-format-lwt.master/opam:35:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/vhd-format.master/opam:30:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/vhd-tool.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/wsproxy.master/opam:7:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-cli-protocol.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-client.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-compression.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-consts.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-datamodel.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-expiry-alerts.master/opam:26:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-log.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-nbd.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-networkd.master/opam:5:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-open-uri.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-rrd-transport-utils.master/opam:5:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-rrd-transport.master/opam:5:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-rrdd.master/opam:38:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-schema.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-storage-cli.master/opam:8:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-tracing.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-types.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-xenopsd-cli.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-xenopsd-simulator.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-xenopsd-xc.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi-xenopsd.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xapi.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xe.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xen-api-client-async.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xen-api-client-lwt.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xen-api-client.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xen-api-sdk.master/opam:9:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/xml-light2.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
packages/xs-extra/zstd.master/opam:6:dev-repo: "git+https://github.com/xapi-project/xen-api.git"
```

[the remark this answers is not in the document]
Comment (lindig) [reply packages/xs-extra/xapi-expiry-alerts.master/opam:26]: Use a URL that works with git. Maybe it worked in the past.

File: packages/xs-extra/xapi.master/opam [hunk 3]
```diff
@@ -51,6 +51,7 @@ depends: [
   "xapi-cli-protocol"
   "xapi-consts"
   "xapi-datamodel"
+  "xapi-expiry-alerts"
```
Comment (gangj) [inline packages/xs-extra/xapi.master/opam:54]: @psafont Thank you for helping to add these missing part!
May I have a question: is this opam used for building the xapi binary: https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/dune#L162 ?
And may I know when will the xapi binary be requested through the opam repo?

File: packages/xs-extra/xapi-expiry-alerts.master/opam (hunk 2 above)
[the remark this answers is not in the document]
Comment (psafont) [reply packages/xs-extra/xapi-expiry-alerts.master/opam:26]: I believe this an opam-specific uri, Python's PIP uses similar URIs to denote git repositories that have to be fetched using HTTPS

File: packages/xs-extra/xapi.master/opam (hunk 3 above)
Comment (psafont) [reply packages/xs-extra/xapi.master/opam:54]: yes, the main xapi binary is part of the xapi opam package. The certificate alert binary is also part of the xapi opam package:
https://github.com/xapi-project/xen-api/blob/master/ocaml/alerts/certificate/dune#L18

**Review by psafont — APPROVED** [review]

File: packages/xs-extra/xapi-expiry-alerts.master/opam (hunk 2 above)
[the remark this answers is not in the document]
Comment (gangj) [reply packages/xs-extra/xapi-expiry-alerts.master/opam:26]: I find a commit which added the "git+" to the `dev-repo` to use Opam 2.0:
```
commit 1802d44dd62aa437ee87f3686394c6c2a821c848
Author: Christian Lindig <christian.lindig@citrix.com>
Date:   Tue Oct 2 14:16:35 2018 +0100

    Update meta data to Opam 2 format
    
    Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

diff --git a/packages/xs-extra/gzip.master/opam b/packages/xs-extra/gzip.master/opam
index a4d07e90..cdc7bf0f 100644
--- a/packages/xs-extra/gzip.master/opam
+++ b/packages/xs-extra/gzip.master/opam
@@ -1,16 +1,24 @@
-opam-version: "1.2"
+opam-version: "2.0"
 maintainer: "xen-api@lists.xen.org"
 authors: "xen-api@lists.xen.org"
 homepage: "https://xapi-project.github.io/"
 bug-reports: "https://github.com/xapi-project/xen-api-libs-transitional.git"
-dev-repo: "https://github.com/xapi-project/xen-api-libs-transitional.git"
-
+dev-repo: "git+https://github.com/xapi-project/xen-api-libs-transitional.git"
 build: [[ "jbuilder" "build" "-p" name "-j" jobs ]]
...
```

Maybe we need to add "git+" for those which missed the "git+" prefix? Like this one:
```
[gangj@xenrt1015818053 xs-opam]$ cat packages/xs-extra/forkexec.master/opam
opam-version: "2.0"
maintainer: "xen-api@lists.xen.org"
authors: "xen-api@lists.xen.org"
homepage: "https://github.com/xapi-project/"
bug-reports: "https://github.com/xapi-project/xen-api/issues"
dev-repo: "https://github.com/xapi-project/xen-api.git"
tags: [ "org:xapi-project" ]

build: [[ "dune" "build" "-p" name "-j" jobs ]]

depends: [
  "ocaml"
  "dune"
  "base-threads"
  "fd-send-recv"
  "ppx_deriving_rpc"
  "rpclib"
  "uuid"
  "xapi-log"
  "xapi-stdext-pervasives"
  "xapi-stdext-unix"
]
synopsis: "Sub-process control service for xapi"
description:
  "This daemon creates and manages sub-processes on behalf of xapi."
url {
  src: "https://github.com/xapi-project/xen-api/archive/master.tar.gz"
}
[gangj@xenrt1015818053 xs-opam]$ 

```

File: packages/xs-extra/xapi.master/opam (hunk 3 above)
Comment (psafont) [reply packages/xs-extra/xapi.master/opam:54]: the opam repository is only used to test packages in github's CI and test install packages in the developers' machines. To change the binaries installed in Xenserver `xapi.spec` must be modified to include the files

Comment (gangj) [reply packages/xs-extra/xapi.master/opam:54]: I see, thank you!

**Pull request merged**